### ISSUE TITLE
Configurable RowKeyFactory.

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#208]
Part of #34. This only introduces a configurable RowKeyFactory, and the logic around instantiating RowKeys.

**Comment by shvachko** [comment]
I created a branch issue-208 for this. I basically carved out RowKeyFactory code from Milan's branch issue-34, made some cleanup and generalized a bit. 
Put back caching. At this point we cache all keys in a single Map. In the future we should have a mapping for each factory, as some of them may be caching, and come not.
Generalized `RowKeyFactory<S>`, where S is the service parameter. FullPathRowKey does not need service, bit INodeIdRowKey will need one.
Removed synchronized in most of the RPC methods that Milan added. Not sure why, they should not be. Some other changes.
Take a look.

**Comment by milandesai** [comment]
Thanks Konstantin, I took a look and the changes looks good. Generalizing RowKeyFactory makes sense to me as well. But we need to ensure that the provided "service" is compatible with the configured RowKeyFactory, or else there will be ClassCastExceptions. I think it may be possible to use additional abstraction here, for example by loading the service and RowKeyFactory together, to enforce this - I'd like to try playing around with this tonight before we commit.

**Comment by milandesai** [comment]
Konstantin, I looked into this some more and I feel uneasy about the way we are using generics here. Some questions/comments:
1. We're doing a lot of unsafe casting so I wanted to be clear on the motives for making RowKeyFactory generic. Is it to allow users to provide their own service implementations for use by the row keys? If so, we should think (in the future) about a configurable provider class that loads the service and the key factory together, thus ensuring the types match.
2. FullPathRowKeyFactory is a RowKeyFactory<Void>, but on creation we cast it to RowKeyFactory<NamespaceAgent>. So the service is expected to be of type Void, but we set it to NamespaceAgent. It won't cause a runtime error since we don't refer to the Void type anywhere in FullPathRowKeyFactory and because getService will return null. So it works, but it's not immediately clear why. We could instead set the type parameter to GiraffaProtocol for now until we design a way to load different service implementations as described previously.
3. Is there a reason why we can't add a field for the service in RowKeyFactory and implement the setter and getter there?
4. We should only load the factory class from the configuration if it has not already been done.

**Comment by shvachko** [comment]
You are right Milan, there is one unnecessary cast, which I eliminated, and you are right we should not ignore the cast exception when we load class from configuration. I'll fix this in the next version. It will be the only cast, you cannot get around this one.
Not sure exactly what you mean in (2). 
We will have the service member in INodeIdRowKeyFactory, with actual setters and getters. FullPathRowKeyFactory just doesn't need a service.
Loading factory only once is a good thing. Let me look into it.

**Comment by shvachko** [comment]
Updated the patch to reflect Milan's comments.
- I essentially left only one unsafe conversion of the factory
- I think I understood what you meant in (2). Not passing any services at all for now.
- Factory is not loaded if it already has been loaded.
- Removed GiraffaProtocol for now. We should introduce it in a separate issue as a separate protocol, rather than extending hdfs.ClientProtocol.

Take a look. Let's get it in and then move forward with GiraffaProtocol next.

**Comment by milandesai** [comment]
Thanks @shvachko. Looks good to me now. +1

**Comment by shvachko** [comment]
Committed. Thanks Milan.
